core/plugins/stack/terraform/version.ts

```diff
@@ -0,0 +1,87 @@
+import { IntrospectFn } from "../../../types.ts";
+
+// Find the latest stable version here:
+// https://www.terraform.io/downloads
+const LATEST = "1.1.2";
+
+const WARN_USING_LATEST =
+  `Couldn't detect the Terraform version, using the latest available: ${LATEST}`;
+
+const ERR_UNDETECTABLE_TITLE =
+  "Couldn't detect which Terraform version this project uses.";
+const ERR_UNDETECTABLE_INSTRUCTIONS = `
+To fix this issue, consider one of the following suggestions:
+
+1. Add the 'required_version' to your .tf file
+
+See https://www.terraform.io/language/settings#specifying-a-required-terraform-version
+
+2. Create a .terraform-version file
+
+The .terraform-version file can be used by the tfenv to choose a specific Terraform version
+for a project.
+
+It's the easiest option, all you have to do is create a .terraform-version text
+file with a version inside, like "1.1.0".
+
+See https://github.com/tfutils/tfenv
+`;
+
+const terraformVerRegex =
+  /required_version.*(?<TerraformVersion>[0-9]+\.[0-9]+\.[0-9])/gm;
+
+function notEmpty<TValue>(value: TValue | null | undefined): value is TValue {
+  return value !== null && value !== undefined;
+}
+
+const tfenv: IntrospectFn<string | Error> = async (context) => {
+  for await (const file of context.files.each("**/.terraform-version")) {
+    return await context.files.readText(file.path);
+  }
+  return Error("Can't find terraform version at .terraform-version");
+};
+
+const tfFile: IntrospectFn<string | Error> = async (context) => {
```

Comment:
Terraform uses a "Version Constraint" string as the `required_version` value. It could be a direct version definition such as `1.1.0` or a range such as `~> 1.0.0` or `>= 1.2.0, < 2.0.0`.

This implementation works for direct version definition and some range definitions, but it could work for any range definition too.

The third example (`>= 1.2.0, < 2.0.0`) would incorrectly return "2.0.0" as the used Terraform version.

In the context object, there are some semantic version helpers. They are useful to detect if the string is a version or a range, and to later extract a version from the defined range. Their signature are defined here:
https://github.com/pipelinit/pipelinit-cli/blob/a3f4cd1/core/types.ts#L31-L40

Can you extend this function to check if the `required_version` is a version or a range and then find the correct version?

See https://www.terraform.io/language/expressions/version-constraints#version-constraint-syntax